libbpf-rs/src/query.rs

```diff
@@ -98,101 +98,305 @@ fn name_arr_to_string(a: &[c_char], default: &str) -> String {
+        Self{opts: opts, ..Self::default()}
+    }
+}
+
+impl ProgInfoQueryOptions {
+    include_field_setter!(include_xlated_prog_insns);
+    include_field_setter!(include_jited_prog_insns);
+    include_field_setter!(include_map_ids);
+    include_field_setter!(include_line_info);
+    include_field_setter!(include_func_info);
+    include_field_setter!(include_jited_line_info);
+    include_field_setter!(include_jited_func_lens);
+    include_field_setter!(include_prog_tags);
+    include_field_setter!(include_jited_ksyms);
+}
+
+macro_rules! update_vec_field {
+    ($item:expr, $flag:expr, $field_and_vec:ident, $len_field:ident, $default:expr, $result_ready:expr) => {
+        if !$flag {
+            $item.$len_field = 0;
+        } else if $field_and_vec.len() != $item.$len_field as usize {
+            $field_and_vec.resize($item.$len_field as usize, $default);
+            $item.$field_and_vec = $field_and_vec.as_mut_ptr() as *mut c_void as u64;
+            $result_ready = false;
+        }
+    };
+}
+
+macro_rules! next_fd {
+    ($cur_id:expr, $next_id_func:expr, $fd_by_id_func:expr) => {
+        loop {
+            if unsafe { $next_id_func($cur_id, &mut $cur_id) } != 0 {
+                return None;
+            }
+
+            let fd = unsafe { $fd_by_id_func($cur_id) };
+            if fd < 0 {
+                if errno::errno() == errno::Errno::ENOENT as i32 {
+                    continue;
+                }
+                return None;
+            }
+            break fd;
+        }
+    };
+}
+
 impl ProgramInfo {
-    fn from_uapi(_fd: i32, s: libbpf_sys::bpf_prog_info) -> Option<Self> {
-        let name = name_arr_to_string(&s.name, "(?)");
-        let ty = match ProgramType::try_from(s.type_) {
-            Ok(ty) => ty,
-            Err(_) => ProgramType::Unknown,
-        };
+    fn load_from_fd(fd: i32, opts: &ProgInfoQueryOptions) -> Option<Self> {
+        // We need to use std::mem::zeroed() instead of just using
+        // ::default() because padding bytes need to be zero as well.
+        // Old kernels which know about fewer fields than we do will
+        // check to make sure every byte past what they know is zero
+        // and will return E2BIG otherwise.
+        let mut item: libbpf_sys::bpf_prog_info = unsafe { std::mem::zeroed() };
+
+        let mut xlated_prog_insns: Vec<u8> = Vec::new();
+        let mut jited_prog_insns: Vec<u8> = Vec::new();
+        let mut map_ids: Vec<u32> = Vec::new();
+        let mut jited_line_info: Vec<*const c_void> = Vec::new();
+        let mut line_info: Vec<LineInfoRaw> = Vec::new();
+        let mut func_info: Vec<FuncInfo> = Vec::new();
+        let mut jited_func_lens: Vec<u32> = Vec::new();
+        let mut prog_tags: Vec<Tag> = Vec::new();
+        let mut jited_ksyms: Vec<*const c_void> = Vec::new();
+
+        let item_ptr: *mut libbpf_sys::bpf_prog_info = &mut item;
+        let mut len = size_of::<libbpf_sys::bpf_prog_info>() as u32;
+
+        loop {
```

Comment:
the actual loop looks unnecessary. You shouldn't need more than two calls for bpf_obj_get_info_by_fd: first to get lengths of necessary buffers, second to fill those buffers out by kernel. And then skip the second step if none of options were set. It should also simplify update_vec_field!() macro, probably, as you know that you are resizing vec.